### ISSUE TITLE
Initializers with the same seed produce different output in Keras 2.7

### ISSUE THREAD

**Issue by drasmuss** [keras-team/keras#15586]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.7.0rc1
- Python version: 3.8
- Bazel version (if compiling from source): N/A
- GPU model and memory: N/A
- Exact command to reproduce:

```
import numpy as np
import tensorflow as tf

init0 = tf.keras.initializers.RandomUniform(seed=0)
w0 = init0((2, 2))

init1 = tf.keras.initializers.RandomUniform(seed=0)
w1 = init1((2, 2))

print(w0)
print(w1)

assert np.allclose(w0, w1)
```

**Describe the problem**.

As far as I understand it, initializers with the same seed should produce the same output. This is how they behaved prior to 2.7. But now passing the same seed to two different initializer instances does not make them produce the same output.

This is likely related to this change mentioned in the TF release notes:
> RNG behavior change for all tf.keras.initializers classes. For any class constructed with a fixed seed, it will no longer generate same value when invoked multiple times. Instead, it will return different value, but a determinisitic sequence. This change will make the initialize behavior align between v1 and v2.

However, that seems to suggest that if I create an initializer with a fixed seed, it will always produce the same sequence of numbers (`x0 -> x1 -> x2 -> ...`). That is not the behaviour observed here; creating two different classes, with the same seed, produces two different sequences. The first class produces `x0 -> x1 -> x2 -> ...`, but the second produces `y0 -> y1 -> y2 -> ...`.

**Describe the current behavior**.

Initializers created with the same seed produce different output.

**Describe the expected behavior**.

Initializers created with the same seed should produce the same output.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): no
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.

https://gist.github.com/drasmuss/212686552bc5f36a7cd23229f9dc4477

**Comment by chunduriv** [comment]
I was able to replicate issues in `TF2.7` and `tf-nightly`. Please find the [gist here](https://colab.research.google.com/gist/chunduriv/efb911b97d1eb2a288afb4ec73db7f3a/15586.ipynb). Thanks!

**Comment by jvishnuvardhan** [comment]
I can reproduce the issue with `tf-nightly`. However, the following workaround may be helpful. [Here](https://colab.research.google.com/gist/jvishnuvardhan/0bbbc0cd98ae79ac135cb3ef0849df1d/15586.ipynb) is a gist for reference. Thanks

```
g = tf.random.Generator.from_seed(0)
w0 = g.normal(shape=(2,2))

g1 = tf.random.Generator.from_seed(0)
w1 = g1.normal(shape=(2,2))

print(w0)
print(w1)
```

```
tf.Tensor(
[[-1.3544159   0.70454913]
 [ 0.03666191  0.86918795]], shape=(2, 2), dtype=float32)
tf.Tensor(
[[-1.3544159   0.70454913]
 [ 0.03666191  0.86918795]], shape=(2, 2), dtype=float32)
```

**Comment by jvishnuvardhan** [comment]
@drasmuss This is working as mentioned in the docs. Please check [here](https://www.tensorflow.org/api_docs/python/tf/keras/initializers/RandomUniform). It was mentioned that the initializer will not produce same random values across multiple calls but produces the same sequence.


seed | A Python integer. Used to create random seeds. See tf.compat.v1.set_random_seed for behavior. Note that seeded initializer will not produce same random values across multiple calls, but multiple initializers will produce same sequence when constructed with same seed value.
-- | --


I ran three times with the seed=0 and verified that the produces the same sequence.  

[Here](https://colab.research.google.com/gist/jvishnuvardhan/c6271cf200f80b5b0e04ede36c2c49c6/15586.ipynb) is a gist for reference. Thanks!

**Comment by drasmuss** [comment]
I don't believe this is working as described in the docs
>Note that seeded initializer will not produce same random values across multiple call

This means that if I construct a given Initializer (e.g. `init0 = tf.keras.initializers.RandomUniform(seed=0)`), and call it multiple times, like
```
init0 = tf.keras.initializers.RandomUniform(seed=0)
w0 = init0((2, 2))
w1 = init0((2, 2))
w2 = init0((2, 2))
```
I should expect `w0`, `w1`, and `w2` to be different. That part is working as expected.

> but multiple initializers will produce same sequence when constructed with same seed value.

This is the part that is not working. This says that multiple initializers, with the same seed value, will produce the same sequence. But that is not the case. If I construct a second initializer, and call it multiple times
```
init1 = tf.keras.initializers.RandomUniform(seed=0)
y0 = init1((2, 2))
y1 = init1((2, 2))
y2 = init1((2, 2))
```
The documentation says that this initializer should produce the same sequence. I.e., `w0 == y0`, `w1 == y1`, and `w2 == y2`. But that is not the case. The two initializers, despite having the same seed value, produce a different sequence.

**Comment by jvishnuvardhan** [comment]
@drasmuss Before running `init1 = tf.keras.initializers.RandomUniform(seed=0)`, if you reset the colab runtime, then the sequence is same (i.e. w0 == y0, w1 == y1, and w2 == y2).

In the above, you are not restarting the runtime. May be we need to update the docs to reflect it. Thanks!

**Comment by drasmuss** [comment]
If you reset the colab runtime then you effectively just have a single initializer (because resetting gets rid of the previous one). The documentation explicitly says that _multiple_ initializers should produce the same sequence with the same seed. You could adjust the documentation to remove that claim entirely, but that's a pretty significant change in the behaviour.

As a broader point, the benefit of setting the seed to a specific value on an object with randomness is to set that randomness to a reliable, reproducible value.  With the current behaviour, the actual seed value passed doesn't really matter, because all the seed does is fix the RNG to some arbitrary point. For example, consider this code snippet:
```
w0 = tf.keras.initializers.RandomUniform(seed=0)((2, 2))
w1= tf.keras.initializers.RandomUniform(seed=0)((2, 2))
w2 = tf.keras.initializers.RandomUniform(seed=1)((2, 2))
```
There is no effective difference here between setting `seed=0` and `seed=1`. In both cases `w0 != w1` and `w0 != w2`. That seems like unexpected behaviour for a seed.

For a more real-world example, consider the case where I want to build a model that has two layers with the same initial weights in each layer. A natural way to go about that would be the following:
```
import tensorflow as tf

my_model = tf.keras.Sequential(
    [
        tf.keras.layers.Dense(
            2,
            kernel_initializer=tf.keras.initializers.RandomUniform(seed=0),
            input_shape=(2,),
        ),
        tf.keras.layers.Dense(
            2,
            kernel_initializer=tf.keras.initializers.RandomUniform(seed=0),
        ),
    ]
)

my_model.build()
print(my_model.weights)
```
I think most users would be pretty surprised to find that that their two layers actually have completely different initial weights, even though they explicitly seeded them to the same value.

**Comment by mattdangerw** [comment]
@qlzh727 could you take a look at this when you have a second?

**Comment by qlzh727** [comment]
Let me add @wangpengmit who is the owner of random ops in tf.

Currently it is using stateful random ops under the hood, which might have some relationship with the tf global seed. In 2.8, we will enable the tf.random.Generator for the initializer, which I think will produce the behavior describe in the docstring.

**Comment by qlzh727** [comment]
Btw, you can take a look for the example in https://www.tensorflow.org/api_docs/python/tf/random/set_seed. There is actually some behavior details. 

In the meantime, you can also try:

```
import numpy as np
import tensorflow as tf

init0 = tf.keras.initializers.RandomUniform(seed=0)
init0._random_generator._force_generator = True
w0 = init0((2, 2))

init1 = tf.keras.initializers.RandomUniform(seed=0)
init1._random_generator._force_generator = True
w1 = init1((2, 2))

print(w0)
print(w1)

assert np.allclose(w0, w1)
```

**Comment by jvishnuvardhan** [comment]
@qlzh727 Depending on the final code change, we also need to update the docs https://www.tensorflow.org/api_docs/python/tf/keras/initializers/RandomUniform

seed | A Python integer. Used to create random seeds. See tf.compat.v1.set_random_seed for behavior. Note that seeded initializer will not produce same random values across multiple calls, but multiple initializers will produce same sequence when constructed with same seed value.
-- | --

**Comment by qlzh727** [comment]
Btw, with the current code, it will also reproduce same result if you have fixed tf.random.seed. It can be removed once have the generator code path turn on by default.

```
import numpy as np
import tensorflow as tf

tf.random.set_seed(1)
init0 = tf.keras.initializers.RandomUniform(seed=0)
w0 = init0((2, 2))

tf.random.set_seed(1)
init1 = tf.keras.initializers.RandomUniform(seed=0)
w1 = init1((2, 2))

print(w0)
print(w1)

assert np.allclose(w0, w1)
```

**Comment by drasmuss** [comment]
I'd wait until TF 2.8 is released to confirm that the behaviour then matches the docs before closing this, but up to you.

**Comment by Lifann** [comment]
I meet same issue here and have to drop some features and unit test to pass the CI checks related to this issue, in TensorFlow 2.7.0

**Comment by wangpengmit** [comment]
This is caused by Keras initializers switching to `tf.random.uniform` in TF 2.7 (as a stop-gap for finally switching to `tf.random.Generator` or `tf.random.stateless_uniform`), which hits one of the weird behaviors of `tf.random.uniform` (the kernel-reusing behavior). As @qlzh727 mentioned, please see https://www.tensorflow.org/api_docs/python/tf/random/set_seed (and search for "kernel" in that page) for detailed explanation.

Calling `set_seed` fixes it because `set_seed` clears all kernel caches.

**Comment by Rabelaiss** [comment]
> Btw, with the current code, it will also reproduce same result if you have fixed tf.random.seed. It can be removed once have the generator code path turn on by default.
> 
> ```
> import numpy as np
> import tensorflow as tf
> 
> tf.random.set_seed(1)
> init0 = tf.keras.initializers.RandomUniform(seed=0)
> w0 = init0((2, 2))
> 
> tf.random.set_seed(1)
> init1 = tf.keras.initializers.RandomUniform(seed=0)
> w1 = init1((2, 2))
> 
> print(w0)
> print(w1)
> 
> assert np.allclose(w0, w1)
> ```

Thank you with this code I was able to get reproducible weights initializaiton in keras models.
I wonder, in order to correctly change the weights initializaiton, is it better to fix the seed in `tf.random.set_seed(1)` and change the seed in `tf.keras.initializers.RandomUniform(seed=0)`, or viceversa?
Many thanks in advance
